lib/common/mock.c

```diff
@@ -92,6 +92,62 @@ __wrap_getenv(const char *name)
 }
 
 
+/* setenv()
+ *
+ * If pcmk__mock_setenv is set to true, later calls to setenv() must be preceded
+ * by:
+ *
+ *     expect_*(__wrap_setenv, name[, ...]);
+ *     expect_*(__wrap_setenv, value[, ...]);
+ *     expect_*(__wrap_setenv, overwrite[, ...]);
+ *     will_return(__wrap_setenv, errno_to_set);
+ *
+ * expect_* functions: https://api.cmocka.org/group__cmocka__param.html
+ *
+ * The mocked function will return 0 if errno_to_set is 0, and -1 otherwise.
+ */
+bool pcmk__mock_setenv = false;
+
+int
+__wrap_setenv(const char *name, const char *value, int overwrite)
+{
+    if (!pcmk__mock_setenv) {
+        return __real_setenv(name, value, overwrite);
+    }
+    check_expected_ptr(name);
+    check_expected_ptr(value);
+    check_expected(overwrite);
+    errno = mock_type(int);
+    return (errno == 0)? 0 : -1;
```

Comment:
This should probably call __real_setenv() for errno == 0 (ditto for unsetenv)

Reply:
I'm on the fence about that.

On the one hand, `__wrap_getenv()` never calls `__real_getenv()` if `pcmk__mock_getenv` is true. Instead it returns the mocked value we gave it via `will_return()`. So it's asymmetrical to have `__wrap_{un,}setenv()` call `__real_{un,}setenv()` if `pcmk__mock_{un,}setenv` is true, even if `errno == 0`. In other words, if the mocked `getenv()` doesn't touch the real environment, you'd think the mocked `setenv()` and `unsetenv()` wouldn't either. That's why I didn't let them produce any side effects besides setting `errno`.

The caller of set/unset also doesn't have to use the "real" results of setting an environment variable since we can mock the return value of `getenv()`.

On the other hand, `__wrap_fopen()` does call `__real_fopen()` if `errno == 0`. Which makes sense since the caller has to actually use the results -- the open file handle. Same with the proposed `__wrap_snprintf()` calling `vsnprintf()` -- we need something in that buffer.